src/coreclr/src/vm/methodtable.h

```diff
@@ -1973,9 +1973,9 @@ class MethodTable
     //
     BOOL CanCastToInterface(MethodTable *pTargetMT, TypeHandlePairList *pVisited = NULL);
     BOOL CanCastToClass(MethodTable *pTargetMT, TypeHandlePairList *pVisited = NULL);
-    BOOL CanCastToClassOrInterface(MethodTable *pTargetMT, TypeHandlePairList *pVisited);
+    BOOL CanCastTo(MethodTable* pTargetMT, TypeHandlePairList *pVisited);
     BOOL ArraySupportsBizarreInterface(MethodTable* pInterfaceMT, TypeHandlePairList* pVisited);
-    BOOL ArrayIsInstanceOf(TypeHandle toTypeHnd, TypeHandlePairList* pVisited);
+    BOOL ArrayIsInstanceOf(MethodTable* pTargetMT, TypeHandlePairList* pVisited);
 
     BOOL CanCastByVarianceToInterfaceOrDelegate(MethodTable* pTargetMT, TypeHandlePairList* pVisited);
 
```

Comment:
Looking through GetGuid, it appears there are paths which reach IsLegalNonArrayWinRTType without an array check. I'm not sure if its possible to actually get there, but it feels like Guid management and this change is a risky combination. (At the very least, I feel there need to be more asserts that invalid paths aren't taken.)

Reply:
I looked through callers to `IsLegalNonArrayWinRTType` and it seems seems to be called only after being sure the type is not an array. 
Some subset of arrays is actually supported in WinRT scenarios and, if relevant, arrays are handled before calling `IsLegalNonArrayWinRTType`.

In the majority of cases the type is known to be an interface or a delegate though.

Reply:
`ReflectionInvocation::GetGUID` does have a bug. Previously it would produce 0000...00 guid for all typedescs. Now it returns a guid for arrays. 

Interestingly, there is also `FCIMPL2(void, MarshalNative::DoGenerateGuidForType` which looks like supposed to work with arrays and produce guids, but I m not sure if there are any callers.

Reply:
> `MarshalNative::DoGenerateGuidForType`

This is dead code. You can delete it while you are on it.